### PR TITLE
UPAPI-22 - Redis client implementation for Authentication module

### PR DESCRIPTION
**Pull request by tpeirotti**

### CODE REVIEW

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+@Service
+public class RedissonCacheServiceImpl implements CacheService {
+
+  private RedissonClient redissonClient;
+  private Long defaultExpirationTimeInSeconds;
+
+  @Autowired
+  public RedissonCacheServiceImpl(
+      RedissonClient redissonClient,
+      @Value("${cache.expirationTimeInSeconds:0}") Long defaultExpirationTimeInSeconds) {
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: Can we have a cache properties and inject that? maybe even selecting the default implementation? similar to email service

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    return defaultExpirationTimeInSeconds;
+  }
+
+  @Override
+  public <T extends Serializable> T get(Module module, final String name, final Class<T> type) {
+    final RBucket<T> bucket = redissonClient.getBucket(name);
+    if (bucket.isExists()) {
+      return bucket.get();
+    }
+    return null;
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: What if we return Optionals?

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  @Override
+  public <T extends Serializable> void put(Module module, final String key, final T obj) {
+    if (defaultExpirationTimeInSeconds > 0) {
+      this.setExpirationInSeconds(module, key, obj, defaultExpirationTimeInSeconds);
+    } else {
+      this.getBucket(module, key).set(obj);
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: can we use in general object or value, etc.

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  }
+
+  @Override
+  public <T extends Serializable> void setExpiration(
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: isn't this like updateExpiration?

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      Module module, final String key, T obj, final long timeValue, final TimeUnit timeUnit) {
+    if (timeValue <= 0 || timeUnit == null) {
+      throw new IllegalArgumentException("Invalid expiry set");
+    }
+    this.getBucket(module, key).set(obj, timeValue, timeUnit);
+  }
+
+  @Override
+  public <T extends Serializable> void setExpirationInSeconds(
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: Same here, isn't updateExpiration

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      final String key = StringUtils.trim(StringUtils.defaultString(s));
+      if (StringUtils.isNotEmpty(key)) {
+        keys.add(buildKey(module, key));
+      }
+    }
+    return redissonClient.getBuckets().get(keys.toArray(new String[0]));
+  }
+
+  @Override
+  public Long remove(Module module, final String... keys) {
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: Can this be delete?

File: src/main/java/com/alsterup/cache/service/CacheService.java [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
-
-  boolean isPresent(Module module, String key);
+    enum Module {
+        Authentication,
+        Role
+    }
+    Long getDefaultExpirationInSeconds();
+    <T extends Serializable> T get(Module module, String name, Class<T> type);
+    <T extends Serializable> void put(Module module, String name, T obj);
+    <T extends Serializable> void setExpiration(Module module, String name, T obj, long timeValue, TimeUnit timeUnit);
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/CacheService.java]: to me this is another `put` not a setExpiration

File: pom.xml [hunk 8]
```diff
@@ -27,6 +27,7 @@
     <opencsv.version>4.6</opencsv.version>
     <coveo.version>2.9</coveo.version>
     <redisson.version>3.11.1</redisson.version>
+    <commons-lang3.version>3.0</commons-lang3.version>
```
Comment (psabbate) [inline pom.xml]: test

Comment (psabbate) [reply pom.xml]: test

Comment (psabbate) [reply pom.xml]: another

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import org.redisson.api.RedissonClient;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Slf4j
+public class RedissonCacheServiceImpl implements CacheService {
+
+  @Autowired private RedissonClient redissonClient;
+  private CacheProperties cacheProperties;
+
+  public RedissonCacheServiceImpl(CacheProperties cacheProperties) {
```
Comment (psabbate) [inline src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: The constructor should be autowired, receiving the redissonClient

File: src/main/java/com/alsterup/security/service/impl/AuthenticationServiceImpl.java [hunk 10]
```diff
@@ -58,25 +60,25 @@
   public AuthenticationServiceImpl(
       SessionService sessionService,
       TokenService tokenService,
+      CacheService defaultRedissonClient,
```
Comment (psabbate) [inline src/main/java/com/alsterup/security/service/impl/AuthenticationServiceImpl.java]: this needs to be the cacheService, not the defaultRedissonClient

File: src/main/resources/application.yaml [hunk 11]
```diff
@@ -32,6 +32,8 @@ spring:
     host: 127.0.0.1
     port: 6379
 
+cache:
+  expirationTimeInSeconds: 0
```
Comment (psabbate) [inline src/main/resources/application.yaml]: I would have expected `defaultExpirationTime` (we assume is seconds) 
and `implementation` and for now having `redis` as default

File: src/test/java/com/alsterup/security/AuthenticationServiceImplTest.java [hunk 12]
```diff
@@ -118,7 +130,7 @@ public void loginTest_successfulLogin() {
             anySet(),
             any(Duration.class),
             anyString());
-    verify(cacheService).put(any(CacheService.Module.class), anyString(), anyString());
+    verify(redissonClient).put(any(CacheService.Module.class), anyString(), anyString());
```
Comment (psabbate) [inline src/test/java/com/alsterup/security/AuthenticationServiceImplTest.java]: don't use redissonClient here

File: pom.xml (hunk 8 above)
Comment (psabbate) [inline pom.xml]: test

Comment (psabbate) [reply pom.xml]: test2

File: src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java (hunk 9 above)
Comment (tpeirotti) [reply src/main/java/com/alsterup/cache/service/impl/RedissonCacheServiceImpl.java]: Beans are wired in CacheConfig class.

File: src/main/java/com/alsterup/cache/configuration/CacheConfig.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties({CacheProperties.class})
+public class CacheConfig {
+
+  private RedissonClient redissonClient;
+
+  @Autowired
+  public CacheConfig(RedissonClient redissonClient) {
```
Comment (pablo-gutierrez-kenility) [inline src/main/java/com/alsterup/cache/configuration/CacheConfig.java]: Esto obliga a tener configurado Reddison en el proyecto, deberia ser por defecto in-memory y que la inyeccion sea a demanda si tengo configurado las props necesarias para levantar el servicio.

File: src/main/java/com/alsterup/security/service/impl/AuthenticationServiceImpl.java [hunk 14]
```diff
@@ -46,7 +47,7 @@
 
   private UserRepository userRepository;
 
-  private CacheService cacheService;
+  private CacheService redissonCacheService;
```
Comment (pablo-gutierrez-kenility) [inline src/main/java/com/alsterup/security/service/impl/AuthenticationServiceImpl.java]: Revertir el nombrado, el servicio no tiene que saber cual es la implementacion del servicio de cache.

File: src/main/java/com/alsterup/security/service/impl/RoleServiceImpl.java [hunk 15]
```diff
@@ -28,12 +27,10 @@
 public class RoleServiceImpl implements RoleService {
 
   private RoleRepository repository;
-  private CacheService cacheService;
 
   @Autowired
-  public RoleServiceImpl(RoleRepository repository, CacheService cacheService) {
```
Comment (pablo-gutierrez-kenility) [inline src/main/java/com/alsterup/security/service/impl/RoleServiceImpl.java]: El servicio de cache deberia mantenerse a este nivel.

File: src/test/java/com/alsterup/security/AuthenticationServiceImplTest.java [hunk 16]
```diff
@@ -36,7 +48,7 @@
 
   @Mock private TokenService tokenService;
   @Mock private UserRepository userRepository;
-  @Mock private CacheService cacheService;
+  @Mock private CacheService redissonCacheService;
```
Comment (pablo-gutierrez-kenility) [inline src/test/java/com/alsterup/security/AuthenticationServiceImplTest.java]: Revertir el nombrado y no usar la implementacion.

File: src/test/java/com/alsterup/security/RoleServiceImplTest.java [hunk 17]
```diff
@@ -22,22 +29,24 @@
...
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
 
 @RunWith(MockitoJUnitRunner.class)
 public class RoleServiceImplTest {
 
   @Mock private RoleRepository repository;
 
-  @Mock private CacheService cacheService;
```
Comment (pablo-gutierrez-kenility) [inline src/test/java/com/alsterup/security/RoleServiceImplTest.java]: El servicio de cache deberia mantenerse a este nivel.

**Pull request merged**